bulk: change AddSSTTable to not be recursive

vivekmenezes opened the pull request:
AddSSTTable was recursive to deal with range splits.
Unfortunately the recursive call would create new SSTs
without freeing the older ones creating a memory buildup
that was quadratic. We've seen memory buildup on the order
of GBs due to this recursion.

fixes #36769
fixes #36381

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36765)
<!-- Reviewable:end -->

vivekmenezes commented:
looks like the tests are failing. Taking a look.

dt commented:
This still forces us to keep making new iterators every time, on top of the post-split SSTs.

I was thinking we might want to make an internal helper wrapped by `AddSSTable` that in addition to the existing args takes an optional iterator. Then we'd pass the iterator, not []byte, to the split helper, so we'd and we could pass the same iterator the whole time. That also would mean that if a given attempt to add failed, we could throw away the []byte result before we attempted to split-and-and, since we could re-iterate the iter rather than iterating the failed file.

vivekmenezes commented:
@dt done as requested. PTAL

dt commented:
if you wanted a test, you could rebase on top of https://github.com/cockroachdb/cockroach/compare/master...dt:addsstable-oom?expand=1

vivekmenezes committed:
bulk: change AddSSTTable to not be recursive

AddSSTTable was recursive to deal with range splits.
Unfortunately the recursive call would create new SSTs
without freeing the older ones creating a memory buildup
that was quadratic. We've seen memory buildup on the order
of GBs due to this recursion.

Release note: None

vivekmenezes commented:
@dt updated this PR with your test. I've confirmed it fails without this PR. PTAL.

dt commented on pkg/storage/bulk/sst_batcher_test.go:
```diff
@@ -203,8 +206,97 @@ func runTestImport(t *testing.T, batchSize int64) {
+			if span.ContainsKey(splits[i]) && !span.Key.Equal(splits[i]) {
+				earlySplit := numKeys / 100
+				if i == earlySplit {
+					early = getMem() - baseline
```
I think we want to get rid of `baseline`  and these `- baseline`s, since they assume that men usage only increases which might not be true -- if for any reason mem usage _dropped_ between establishing baseline once of these calls, this subtraction just overflows.

dt commented on pkg/storage/bulk/sst_batcher.go:
```diff
@@ -189,57 +189,99 @@ func (b *SSTBatcher) GetSummary() roachpb.BulkOpSummary {
+func createSplitSSTable(
+	ctx context.Context,
+	db sender,
+	sstBytes []byte,
```
I don't think we need to pass `sstBytes` and `iter` -- I think we'd be okay with just `iter`?

vivekmenezes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36765)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/storage/bulk/sst_batcher.go, line 267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36765#-LcC0SBVY4Eez5Q-4tcQ-r1-267:-LcI9dGk9aLkfOj2c05j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17beae792f05a2482419746a35754fe84b8defb9/pkg/storage/bulk/sst_batcher.go#L267)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I don't think we need to pass `sstBytes` and `iter` -- I think we'd be okay with just `iter`?
</blockquote></details>

Done.
___
*[pkg/storage/bulk/sst_batcher_test.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36765#-LcCan3LxBi49GU8u_3g-r1-279:-LcI8NfR2n9KEFqsTfhm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17beae792f05a2482419746a35754fe84b8defb9/pkg/storage/bulk/sst_batcher_test.go#L279)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we want to get rid of `baseline`  and these `- baseline`s, since they assume that men usage only increases which might not be true -- if for any reason mem usage _dropped_ between establishing baseline once of these calls, this subtraction just overflows.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=36765) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=36765) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=36765) it.</sub>

vivekmenezes commented:
bors r+

On Sat, 13 Apr 2019, 8:31 AM David Taylor, <notifications@github.com> wrote:

> *@dt* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/36765#pullrequestreview-226351380>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBKEOTMtOgeEVUtcGKnhOgxcTq8GUks5vgc4qgaJpZM4cp5ZV>
> .
>

vivekmenezes commented:
bors r+
